File: signosoft_signer/ios/signosoft_signer/Sources/signosoft_signer/SignosoftSignerPlugin.swift
import Flutter
import UIKit

// Under Swift Package Manager the core is a separate module. Under CocoaPods
// the podspec compiles it into this same module, so there is nothing to import.
#if canImport(SignosoftSigner)
import SignosoftSigner
#endif

/// Bridges the Dart `SignosoftSigner` API onto `SignosoftSigner.present`.
public final class SignosoftSignerPlugin: NSObject, FlutterPlugin {
    private static let channelName = "com.signosoft.signer"

    private let channel: FlutterMethodChannel
    private var isPresenting = false

    init(channel: FlutterMethodChannel) {
        self.channel = channel
        super.init()
    }

    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(
            name: channelName,
            binaryMessenger: registrar.messenger()
        )
        registrar.addMethodCallDelegate(
            SignosoftSignerPlugin(channel: channel),
            channel: channel
        )
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard call.method == "open" else {
            result(FlutterMethodNotImplemented)
            return
        }

        let arguments = call.arguments as? [String: Any] ?? [:]
        guard let token = arguments["token"] as? String, !token.isEmpty else {
            result(Self.error(.invalidToken, "A bioid token is required."))
            return
        }
        guard let baseUrlString = arguments["baseUrl"] as? String,
              let baseURL = URL(string: baseUrlString)
        else {
            result(Self.error(.invalidBaseUrl, "A valid baseUrl is required."))
            return
        }
        guard let presenter = Self.topViewController() else {
            result(Self.error(.noPresenter, "No view controller available to present the signer."))
            return
        }
        guard !isPresenting else {
            result(Self.error(.alreadyOpen, "A signing session is already open."))
            return
        }

        isPresenting = true
        SignosoftSigner.present(
            from: presenter,
            token: token,
            baseURL: baseURL,
            loadTimeout: Self.loadTimeout(arguments["loadTimeoutMs"]),
            onEvent: Self.wantsDiagnostics(arguments) ? { [weak self] event, data in
                self?.channel.invokeMethod(
                    "diagnostic",
                    arguments: ["event": event, "data": data]
                )
            } : nil
        ) { [weak self] outcome in
            self?.isPresenting = false
            switch outcome {
            case .signed(let info):
                result(Self.payload(status: "signed", info: info))
            case .rejected(let info):
                result(Self.payload(status: "rejected", info: info))
            case .cancelled:
                result(["status": "cancelled"])
            case .error(let error):
                let signosoft = error as? SignosoftError
                result(Self.error(
                    signosoft?.code ?? .unknown,
                    signosoft?.message ?? error.localizedDescription
                ))
            }
        }
    }

    private static func payload(status: String, info: SignedInfo) -> [String: Any?] {
        [
            "status": status,
            "result": info.result,
            "document": info.document,
            "documentToken": info.documentToken,
            "lang": info.lang,
            "signaturesSigned": info.signaturesSigned,
            "signaturesTotal": info.signaturesTotal,
            "lastSignerFirstName": info.lastSignerFirstName,
            "lastSignerLastName": info.lastSignerLastName,
            "lastSignerEmail": info.lastSignerEmail,
            "downloadUrl": info.downloadUrl?.absoluteString,
            "signedPdfPath": info.signedPdfFileURL?.path,
        ]
    }

    /// Dart branches on `details["code"]`; the message is only ever shown to a
    /// developer.
    private static func error(_ code: SignosoftErrorCode, _ message: String) -> FlutterError {
        FlutterError(
            code: "signosoft_error",
            message: message,
            details: ["code": code.rawValue]
        )
    }

    private static func loadTimeout(_ value: Any?) -> TimeInterval {
        guard let milliseconds = (value as? NSNumber)?.doubleValue, milliseconds > 0 else {
            return SignosoftSignerViewController.defaultLoadTimeout
        }
        return milliseconds / 1000
    }

    /// Diagnostics cost a channel round trip per bridge event, so they are only
    /// forwarded when Dart asked for them.
    private static func wantsDiagnostics(_ arguments: [String: Any]) -> Bool {
        arguments["diagnostics"] as? Bool ?? false
    }

    /// Works whether or not the host app uses a scene delegate, so there is no
    /// AppDelegate window to reach for.
    private static func topViewController() -> UIViewController? {
        let window = UIApplication.shared.connectedScenes
            .compactMap { $0 as? UIWindowScene }
            .flatMap(\.windows)
            .first { $0.isKeyWindow }

        var controller = window?.rootViewController
        while let presented = controller?.presentedViewController {
            controller = presented
        }
        return controller
    }
}

File: signosoft_signer/android/src/main/kotlin/com/signosoft/signer/flutter/SignosoftSignerPlugin.kt
package com.signosoft.signer.flutter

import android.app.Activity
import android.content.Intent
import android.net.Uri
import com.signosoft.signer.SignedInfo
import com.signosoft.signer.SignosoftErrorCode
import com.signosoft.signer.SignosoftSigner
import com.signosoft.signer.SignosoftSignerRequest
import com.signosoft.signer.SignosoftSignerResult
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.PluginRegistry.ActivityResultListener

/** Bridges the Dart `SignosoftSigner` API onto [SignosoftSigner]. */
class SignosoftSignerPlugin :
    FlutterPlugin,
    MethodChannel.MethodCallHandler,
    ActivityAware,
    ActivityResultListener {

    private lateinit var channel: MethodChannel
    private var binding: ActivityPluginBinding? = null
    private var pending: MethodChannel.Result? = null

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, CHANNEL_NAME)
        channel.setMethodCallHandler(this)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        this.binding = binding
        binding.addActivityResultListener(this)
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) =
        onAttachedToActivity(binding)

    override fun onDetachedFromActivityForConfigChanges() = onDetachedFromActivity()

    override fun onDetachedFromActivity() {
        binding?.removeActivityResultListener(this)
        binding = null
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        if (call.method != "open") {
            result.notImplemented()
            return
        }

        val token = call.argument<String>("token")
        if (token.isNullOrEmpty()) {
            result.fail(SignosoftErrorCode.INVALID_TOKEN, "A bioid token is required.")
            return
        }
        val baseUrl = call.argument<String>("baseUrl")?.let(Uri::parse)
        if (baseUrl == null || !SignosoftSigner.isUsableBaseUrl(baseUrl)) {
            result.fail(SignosoftErrorCode.INVALID_BASE_URL, "A valid baseUrl is required.")
            return
        }
        val activity = binding?.activity
        if (activity == null) {
            result.fail(SignosoftErrorCode.NO_PRESENTER, "No activity available to open the signer.")
            return
        }
        if (pending != null) {
            result.fail(SignosoftErrorCode.ALREADY_OPEN, "A signing session is already open.")
            return
        }

        pending = result
        val request = SignosoftSignerRequest(
            token = token,
            baseUrl = baseUrl,
            loadTimeoutMillis = loadTimeoutOf(call.argument<Number>("loadTimeoutMs")),
            // Diagnostics cost a channel round trip per bridge event, so they
            // are only forwarded when Dart asked for them.
            onDiagnostic = if (call.argument<Boolean>("diagnostics") == true) {
                { event, data -> channel.invokeMethod("diagnostic", mapOf("event" to event, "data" to data)) }
            } else {
                null
            },
        )
        activity.startActivityForResult(
            SignosoftSigner.createIntent(activity, request),
            REQUEST_CODE,
        )
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?): Boolean {
        if (requestCode != REQUEST_CODE) return false
        val result = pending ?: return true
        pending = null

        when (val outcome = SignosoftSigner.parseResult(resultCode, data)) {
            is SignosoftSignerResult.Signed -> result.success(payloadFor("signed", outcome.info))
            is SignosoftSignerResult.Rejected -> result.success(payloadFor("rejected", outcome.info))
            SignosoftSignerResult.Cancelled -> result.success(mapOf("status" to "cancelled"))
            is SignosoftSignerResult.Failed -> result.fail(outcome.code, outcome.message)
        }
        return true
    }

    /**
     * Dart branches on `details["code"]`; the message is only ever shown to a
     * developer.
     */
    private fun MethodChannel.Result.fail(code: SignosoftErrorCode, message: String) =
        error("signosoft_error", message, mapOf("code" to code.wire))

    private companion object {
        const val CHANNEL_NAME = "com.signosoft.signer"
        const val REQUEST_CODE = 0x5165
    }
}

/** A missing or nonsensical timeout falls back to the core's default. */
internal fun loadTimeoutOf(value: Number?): Long {
    val milliseconds = value?.toLong() ?: 0
    return if (milliseconds > 0) milliseconds else SignosoftSigner.DEFAULT_LOAD_TIMEOUT_MILLIS
}

/** The reply Dart's `parseSignerResult` reads. Keys are shared with iOS. */
internal fun payloadFor(status: String, info: SignedInfo): Map<String, Any?> = mapOf(
    "status" to status,
    "result" to info.result,
    "document" to info.document,
    "documentToken" to info.documentToken,
    "lang" to info.lang,
    "signaturesSigned" to info.signaturesSigned,
    "signaturesTotal" to info.signaturesTotal,
    "lastSignerFirstName" to info.lastSignerFirstName,
    "lastSignerLastName" to info.lastSignerLastName,
    "lastSignerEmail" to info.lastSignerEmail,
    "downloadUrl" to info.downloadUrl,
    "signedPdfPath" to info.signedPdfFile?.path,
)
